source/tricky-pipe/src/tp3.rs

```diff
@@ -0,0 +1,1071 @@
+use crate::loom::{
+    cell::{self, UnsafeCell},
+    sync::atomic::{self, AtomicUsize, Ordering::*},
+};
+use core::{
+    cmp, fmt,
+    mem::{ManuallyDrop, MaybeUninit},
+    ops::{Deref, DerefMut},
+};
+use maitake_sync::{WaitCell, WaitQueue};
+use mnemos_bitslab::index::IndexAllocWord;
+use serde::{de::DeserializeOwned, Serialize};
+
+const CAPACITY: usize = IndexAllocWord::CAPACITY as usize;
+const SHIFT: usize = CAPACITY.trailing_zeros() as usize;
+const SEQ_ONE: usize = 1 << SHIFT;
+const MASK: usize = SEQ_ONE - 1;
+
+#[cfg(not(test))]
+macro_rules! dbg {
+    ($x:expr) => {
+        $x
+    };
+}
+
+pub struct TrickyPipe<T> {
+    elements: [UnsafeCell<MaybeUninit<T>>; CAPACITY],
+    core: Core,
+}
+
+struct Core {
+    dequeue_pos: AtomicUsize,
+    enqueue_pos: AtomicUsize,
+    cons_wait: WaitCell,
+    prod_wait: WaitQueue,
+    indices: IndexAllocWord,
+    queue: [AtomicUsize; CAPACITY],
```

Comment:
these could be `AtomicU16`s to save space. with a 6-bit index (0-63), we need 6 bits of slot index and 6 bits of stored element index, which gives us a 4-bit sequence counter (see https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a463fe344759b479c2fd680984df854c), which seems fine?